pkg/collector/corechecks/snmp/internal/report/report_bandwidth_usage.go

```diff
@@ -97,18 +98,76 @@ func (ms *MetricSender) sendBandwidthUsageMetric(symbol profiledefinition.Symbol
 	}
 	usageValue := ((octetsFloatValue * 8) / (float64(ifSpeed))) * 100.0
 
-	sample := MetricSample{
-		value:      valuestore.ResultValue{SubmissionType: profiledefinition.ProfileMetricTypeCounter, Value: usageValue},
-		tags:       tags,
-		symbol:     profiledefinition.SymbolConfig{Name: usageName + ".rate"},
-		forcedType: profiledefinition.ProfileMetricTypeCounter,
-		options:    profiledefinition.MetricsConfigOption{},
+	interfaceID := ms.deviceID + ":" + fullIndex + "." + usageName
+	err = ms.calculateRate(interfaceID, ifSpeed, usageValue, usageName, tags)
+	if err != nil {
+		return err
 	}
+	return nil
+}
+
+/*
+calculateRate is responsible for checking the state for previously seen metric sample to generate the rate from.
+If the ifSpeed has changed for the interface, the rate will not be submitted (drop the previous sample)
+*/
+func (ms *MetricSender) calculateRate(interfaceID string, ifSpeed uint64, usageValue float64, usageName string, tags []string) error {
+	ms.interfaceRateMap.mu.RLock()
+	interfaceRate, ok := ms.interfaceRateMap.rates[interfaceID]
+	ms.interfaceRateMap.mu.RUnlock()
+
+	// current data point has the same interface speed as last data point
+	if ok && interfaceRate.ifSpeed == ifSpeed {
+		// calculate the delta, taken from pkg/metrics/rate.go
+		// https://github.com/DataDog/datadog-agent/blob/ecedf4648f41193988b4727fc6f893a0dfc3991e/pkg/metrics/rate.go#L37
+		currentTimestamp := TimeNow()
+		delta := (usageValue - interfaceRate.previousSample) / (currentTimestamp - interfaceRate.previousTs)
+		// update the map previous as the current for next rate
+		interfaceRate.previousSample = usageValue
+		interfaceRate.previousTs = currentTimestamp
```

Comment:
Is there a reason for storing the timestamp as float64 instead of `int64` (type of `time.now().UnixNano()`) ? It seems cleaner to store the original int64 format rather than a modified value (float64).

Maybe we can use timestamp as int64 (`UnixNano()` output) and then convert to float when/if needed.

Reply:
i see, if i'm understanding correctly: defer the `/ float64(time.Second)` until i need to use it when i'm calculating the rate / storing the previous value in the map - i can do that!

i'll probably rename the field to `timestampNano` to make it more clear what it is